File: Medium/longest_palindrome_substring.go
package main

import (
	"fmt"
)

func longestPalindrome(s string) string {
	if len(s) == 0 || len(s) > 1000 {
		return s
	}

	if s == reverseString(s) {
		return s
	}

	var resultMatch string
	strSplit := []rune(s)
	lenStr := len(strSplit)
	resultMatch = string(strSplit[0])
	for i := 0; i < lenStr; i++ {
		for j := i + 1; j <= lenStr; j++ {
			if strSplit[i] != strSplit[j-1] {
				continue
			}

			if j > 1 && i+1 < lenStr && strSplit[i+1] != strSplit[j-2] {
				continue
			}

			if len(resultMatch) >= j-i {
				continue
			}

			strJoin := string(strSplit[i:j])
			strJoinReverse := reverseString(strJoin)

			// if i == 1 {
			// fmt.Printf("%d - %d \n", i, j)
			// }

			if strJoin == strJoinReverse {
				resultMatch = strJoin
			}
		}
	}

	return resultMatch
}

func run_longestPalindrome() {
	// start := time.Now()
	fmt.Println(longestPalindrome("cbbd"))
	fmt.Println(longestPalindrome("babad"))
	fmt.Println(longestPalindrome("ccc"))
	// d := time.Since(start)
	// fmt.Println(d)
}

File: Medium/longest_palindrome_substring.js
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function (s) {
  if (s.length == 0 || s.length > 1000) {
    return s;
  }

  if (s == s.split("").reverse().join("")) {
    return s;
  }

  var resultMatch = s[0];
  for (let i = 0; i < s.length; i++) {
    for (let j = i + 1; j <= s.length; j++) {
      if (s[i] != s[j - 1]) {
        continue;
      }

      if (j > 1 && i + 1 < s.length && s[i + 1] != s[j - 2]) {
        continue;
      }

      if (resultMatch.length >= j - i) {
        continue;
      }

      let strJoin = s.slice(i, j);
      let strJoinReverse = strJoin.split("").reverse().join("");

      if (strJoin == strJoinReverse) {
        resultMatch = strJoin;
      }
    }
  }

  return resultMatch;
};

console.log(longestPalindrome("cbbd"));
console.log(longestPalindrome("babad"));
console.log(longestPalindrome("ccd"));
